### ISSUE TITLE
Importing numba takes almost half a second

### ISSUE THREAD

**Issue by Horta** [numba/numba#2550]
```
> time python -c "import numba"
python -c "import numba"  0.40s user 0.08s system 90% cpu 0.532 total
```

**Comment by seibert** [comment]
Yeah, thanks for the reminder.  This time tends to slowly creep up, and we have to periodically make some changes to stomp it back down again.  We'll take a look.

**Comment by fabioz** [comment]
I see that `numba.__init__` imports quite a lot... ideally, my suggestion would be instead of doing things as:

`from .special import typeof`

do something as:

```
def typeof(*args, **kwargs):
    from .special import typeof
    return typeof(*args, **kwargs)
```

-- in many places... 

as well as changing for local imports on some other places...

i.e.: I think there's really no magic, importing in Python means executing lots of code to create classes/methods at runtime, so, the only way to actually make importing faster is delaying that execution.

If such a pull request was done, would it be accepted?  

p.s.: I don't really know when/ if I'll be able to get time to do this, but I'm curious to know if such an approach would be accepted anyways before even considering doing it.
